src/rgw/rgw_client_io.h

```diff
@@ -4,83 +4,236 @@
 #ifndef CEPH_RGW_CLIENT_IO_H
 #define CEPH_RGW_CLIENT_IO_H
 
+#include <exception>
+#include <string>
 #include <streambuf>
 #include <istream>
 #include <stdlib.h>
 
-#include "include/types.h"
+#include <boost/utility/string_ref.hpp>
 
+#include "include/types.h"
 #include "rgw_common.h"
 
 class RGWClientIO {
-  bool _account;
-
 protected:
-  RGWEnv env;
-
   virtual void init_env(CephContext *cct) = 0;
 
 public:
   virtual ~RGWClientIO() {}
-  RGWClientIO() : _account(false) {}
 
   void init(CephContext *cct);
-  RGWEnv& get_env() { return env; }
+  virtual RGWEnv& get_env() noexcept = 0;
+  virtual int complete_request() = 0;
+}; /* RGWClient IO */
+
+
+class RGWClientIOAccounter {
+public:
+  virtual ~RGWClientIOAccounter() {}
+
+  virtual void set_account(bool enabled) = 0;
+
+  virtual uint64_t get_bytes_sent() const = 0;
+  virtual uint64_t get_bytes_received() const = 0;
+};
+
+
+class RGWRestfulIOEngine : public RGWClientIO {
+  template<typename T> friend class RGWDecoratedRestfulIO;
+
+public:
+  class Exception : public std::exception {
+    int err;
+  public:
+    Exception(const int err)
+      : err(err) {
+    }
 
-  bool account() { return _account; }
-  void set_account(bool _accnt) {
-    _account = _accnt;
+    int value() {
+      return err;
+    }
+  };
+
+  virtual std::size_t send_status(int status, const char *status_name) = 0;
```

Comment:
This is because in C++, unlike in C, [`size_t` lives in the `std` namespace](http://en.cppreference.com/w/cpp/types/size_t). Although `std::size_t` looks really ugly, I would prefer to avoid putting global `using` in headers.